### PR TITLE
Refactor HTTP client/Ratelimiter

### PR DESCRIPTION
**Pull request by strideynet**
I've tidied up the overlap of concerns a fair bit here, meaning that the ratelimiter is now no longer responsible for setting auth headers etc (and so that code is no longer duplicated). 

As it stands, I'll be using my fork of the library, but I'm willing to accommodate some change requests if that means this change can be merged upstream.

I've also made a few other decisions here in an attempt to make stuff simpler:
- Renamed redis and memory go files to include "ratelimit". This groups them next to the ratelimit.go file and also makes it clearer its related functionality.
- I've removed the `requestNoAuth` method and replaced it with a field on the request type. This means less ground to cover in terms of testing, and less duplication of code.

For now this is a draft, I won't do the refactor to the Redis ratelimiter until this has had some eyes on it.

### CODE REVIEW

**Review by JakeMakesStuff — APPROVED** [review]
I like the look of this so far

**Comment by strideynet** [comment]
Cool. I've finished the test bits I wanted to do before I could use it. I'll await @Kelwing's review before I refactor the Redis rate limiter.

**Review by Kelwing — COMMENTED** [review]
lgtm

**Review by Kelwing — APPROVED** [review]

File: client.go
Comment (Jack073) [reply client.go:13]: I'd say Authorization makes more sense here actually yep, partly for consistency and because it's not technically just the token.

**Review by Jack073 — COMMENTED** [review]
LGTM (just the Config.Token -> Config.Authorization you spotted)

**Comment by Kelwing** [comment]
I'll merge this once Jack puts his final stamp of approval on it.

**Review by Jack073 — APPROVED** [review]

**Pull request merged**
